Realm/RLMQueryUtil.mm

```diff
@@ -126,13 +126,62 @@ void set_table() override {}
     return [NSString stringWithFormat:@"unknown operator %lu", (unsigned long)operatorType];
 }
 
+class CollectionOperation {
+public:
+    enum Type {
+        Count,
+    };
+
+    CollectionOperation(Type type, RLMProperty *linkProperty, RLMProperty *property)
+        : m_type(type)
+        , m_link_property(linkProperty)
+        , m_property(property)
+    {
+        RLMPrecondition(m_link_property.type == RLMPropertyTypeArray, @"Invalid predicate", @"Collection operation can only be applied to a property of type RLMArray.");
+        switch (m_type) {
+            case Count:
+                RLMPrecondition(!m_property, @"Invalid predicate", @"Result of @count does not have any properties.");
+                break;
+        }
+    }
+
+    CollectionOperation(NSString *operationName, RLMProperty *linkProperty, RLMProperty *property = nil)
+        : CollectionOperation(type_for_name(operationName), linkProperty, property)
+    {
+    }
+
+    Type type() const { return m_type; }
+    NSUInteger linkColumnIndex() const { return m_link_property.column; }
+    NSUInteger columnIndex() const { return m_property.column; }
+
+    void validate_value(id value) const {
+        switch (m_type) {
+            case Count:
+                RLMPrecondition([value isKindOfClass:[NSNumber class]], @"Invalid operand", @"@count can only be compared with a numeric value.");
```

Comment:
Key path vs key path comparisons need more work in general since they are currently limited to single level keys. I'm hoping we'll be able to unify their handling with key path vs value comparisons since it'd be nice for them to support the same operators and to avoid writing similar code in multiple places when adding new functionality.